File: testdata/fixtures/ruby/safe/command_safe.rb
# Safe: Command execution with array form (no shell interpolation)
# Expected: No findings for GTSS-INJ-002

require 'open3'

class FileController < ApplicationController
  def convert
    input_path = params[:path]
    # Safe: Open3.capture3 with array arguments prevents shell injection
    stdout, stderr, status = Open3.capture3("convert", input_path, "-resize", "100x100", "/tmp/thumb.png")
    if status.success?
      send_file "/tmp/thumb.png"
    else
      render plain: "Conversion failed", status: 500
    end
  end

  def list_files
    directory = params[:dir]
    # Safe: Array form of system() call
    stdout, stderr, status = Open3.capture3("ls", "-la", directory)
    render plain: stdout
  end

  def safe_ping
    host = params[:host]
    # Validate input before use
    unless host.match?(/\A[\w.\-]+\z/)
      render plain: "Invalid host", status: 400
      return
    end
    stdout, stderr, status = Open3.capture3("ping", "-c", "3", host)
    render plain: stdout
  end
end

File: testdata/fixtures/python/safe/command_safe.py
# Safe: subprocess.run with list arguments and shell=False
# Expected: No findings for GTSS-INJ-002

import subprocess
import shlex
from flask import Flask, request

app = Flask(__name__)

ALLOWED_HOSTS = {"example.com", "test.internal"}


@app.route("/ping")
def ping():
    host = request.args.get("host", "")
    if host not in ALLOWED_HOSTS:
        return "Invalid host", 400
    result = subprocess.run(
        ["ping", "-c", "3", host],
        capture_output=True,
        text=True,
        timeout=10,
    )
    return result.stdout


@app.route("/convert")
def convert():
    filename = request.args.get("file", "")
    if not filename.isalnum():
        return "Invalid filename", 400
    result = subprocess.run(
        ["convert", f"/uploads/{filename}", "-resize", "100x100", "/tmp/out.png"],
        capture_output=True,
        text=True,
    )
    return "Converted" if result.returncode == 0 else "Failed"


@app.route("/disk-usage")
def disk_usage():
    result = subprocess.run(
        ["df", "-h"],
        capture_output=True,
        text=True,
    )
    return result.stdout
